### DISCUSSION TITLE
How I can track the stock of products quantity while purchasing and selling the products- I will pay for it.

### DISCUSSION THREAD

**Discussion by imhemayatsangin** [serenity-is/Serenity#6407] [Help]
Greetings members,
I am working on hospital management system and in the pharmacy section I want to track the product qty stock. While purchasing the products and also selling the products quantities should increase and decrease in the stock with the purchase and selling behaviors.

How to do it I tried but I failed.
 If anybody help want to help me for free I will be grateful to them or else I am going to pay for it. Please send me an email and I will give you full details. 
Hemayat.sangin@gmail.com
Regards,
Sangin

**Comment by VictorTomaili** [comment]
You can get “premium” and/or “additional support” from serenity team by sending email to sales. Check [serenity.is](https://serenity.is) to contact.

**Comment by stixoffire** [comment]
You can easily do this in your SQL Database by using a trigger - you can also do it from your code - I implemented this in my SQL using a trigger - when the order was completed - I have a record of that order and the items / quantities of the order.
So then for each item on the order - I would simply subtract those quantities from my item stock.
Simple as that .. 



There are two approaches to use (each has their own pros and cons) -
Method 1:
After order complete - you could have code execute there and subtract quantities of the order that was placed from your stock items. 

Method 2: 
SQL trigger on your orders database - when the order has been placed a trigger should fire and then from the inserted order (or even an updated order that was placed erroneously for example I took 3 but fat fingered a 6 on keys ) .

Pros :  Method 1 -  All your code is in one place ..
Cons : Method 1 - if the web session gets disconnected before the code is able to execute on the database - your stock quantity might not get updated.

Pros :  Method 2 -  When the order is placed it is always updated by the database - simpler to handle updates of the order .
Cons:  Method 2 -  Your code is in more than one place .. updating and managing the code base becomes a little more involved.


I used Method 2 and my reasoning was I did not want to have a disconnected session messing up my counts and I also wanted to be able to handle the updates.

In MS SQL the trigger would query would be on **After Insert** and to handle Updates On **After Update**   


```
UPDATE 
    MyStockItemTable
SET 
    MyStockItemTable.Qty = (MyStockItemTable.Qty - Inserted.Qty)
   -- for anything else you want to update each row MyStockItemTable.c2 = expression,
    ...   
FROM 
    MyStockItemTable
    [INNER | LEFT] JOIN Inserted ON MyStockItemTable.ItemId = Inserted.ItemId
WHERE 
    where_predicate; -- You may not even need a where clause .. the join should handle it .
```




For after updated you _Might_ need to create a temp table in your query of the original values of the order before update , 
In MS SQL you can do something like this in the expression of setting the value : MyStockItemTable.Qty = MyStockItemTable.Qty  - (Updated.Qty - Original.Qty ) ;
MyStockItemTable.Qty = 108
 Updated.Qty = 6 , Original Qty = 4 ,
 Formula [108-(6-4)] > [108-2] = 106
 
MyStockItemTable.Qty = 108
 Updated.Qty =3 , Original Qty = 9,
 Formula [108-(3-9)] > [108-(-6)] = 114
 
 

Inserted / Original is the items from your order list that were inserted into the order (OrderItems table).


 If you need more .. see my email to you.
